### PR TITLE
adding 3D structure Mapping Vis to LigandMapping

### PR DESCRIPTION
**Pull request by RiesBen**
Here the visualization functions are added in order to visualize the mapping in 3D.

new dependency: py3DMol

New feature:
```
lomap_mapper = LomapAtomMapper()
lomap_mapping = next(lomap_mapper.suggest_mappings(SmallMoleculeComponent(molA) , SmallMoleculeComponent(molB)))
lomap_mapping.show_mapping3D()

```
image

### CODE REVIEW

File: openfe/setup/atom_mapping/ligandatommapping.py [hunk 1]
```diff
@@ -2,12 +2,15 @@
 # For details, see https://github.com/OpenFreeEnergy/openfe
 import json
 import gufe
-from typing import Any, Optional
 import numpy as np
+from typing import Any, Optional
 from numpy.typing import NDArray
 
+from py3Dmol import view
```
Comment (IAlibay) [inline openfe/setup/atom_mapping/ligandatommapping.py:9]: We'll probably want to make this an optional dependency like perses. py3dmol can be quite heavy with all the node.js things

File: openfe/setup/atom_mapping/ligandatommapping.py [hunk 2]
```diff
@@ -147,6 +150,23 @@ def draw_to_file(self, fname: str, d2d=None):
...
+        ----------
+        show_atomIDs : bool, optional
+            show the atomIDs of the Molecules, by default True
+
+        Returns
+        -------
+        view
+            the visualization of the mapping on the 3D structure of the components, can be further adjusted
+        """
+        view = draw_mapping_on_3Dstructure(self, show_atomIDs=show_atomIDs)
```
Comment (IAlibay) [inline openfe/setup/atom_mapping/ligandatommapping.py:167]: Was draw_mapping_3d not long enough :P

**Review by IAlibay — CHANGES_REQUESTED** [review]
Just the one comment re: optional deps. Do we have tests for these visualization methods? Might be worth adding one if possible, at the very least something that flushes the code path will tell us if things break horribly.

**Comment by IAlibay** [comment]
o also mypy is annoyed for some reason

File: openfe/setup/atom_mapping/ligandatommapping.py (hunk 2 above)
Comment (RiesBen) [reply openfe/setup/atom_mapping/ligandatommapping.py:167]: hihi

**Comment by RiesBen** [comment]
> Just the one comment re: optional deps. Do we have tests for these visualization methods? Might be worth adding one if possible, at the very least something that flushes the code path will tell us if things break horribly.

added a smoke test for this... hm not very nice but maybe sufficient?

**Review by richardjgowers — CHANGES_REQUESTED** [review]
this might need a rework due to some refactoring I'm about to do

**Comment by RiesBen** [comment]
@richardjgowers HA, now I want to move this to kartograf! :P what do you think? :)

**Comment by richardjgowers** [comment]
@RiesBen nah this is useful for *all* mapping objects, so it's openfe :P

**Comment by RiesBen** [comment]
I tried :)

File: openfe/utils/visualization_3D.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+                "center": {"x": p2.x, "y": p2.y, "z": p2.z},
+                "radius": 0.6,
+                "color": color,
+                "alpha": 0.8,
+            }
+        )
+
+
+@requires_package("py3Dmol")
+def draw_mapping_on_3Dstructure(
```
Comment (richardjgowers) [inline openfe/utils/visualization_3D.py:111]: the only thing I don't like is the name... `draw_3D_mapping`?  `show_3D_mapping`?

**Review by richardjgowers — DISMISSED** [review]

File: openfe/utils/visualization_3D.py (hunk 3 above)
Comment (RiesBen) [reply openfe/utils/visualization_3D.py:111]: hehe ok now you and Irfan are two :) let's go with show_3D_mapping. :)

**Review by richardjgowers — APPROVED** [review]

**Pull request merged**